File: app/src/main/java/com/estebanlamas/planesradar/domain/UseCase.java
package com.estebanlamas.planesradar.domain;

import com.estebanlamas.planesradar.domain.executor.PostExecutionThread;

import rx.Observable;
import rx.Scheduler;

/**
 * Created by esteban on 26/12/17
 */

public abstract class UseCase<T> {
    private final Scheduler threadExecutor;
    private final PostExecutionThread mainThread;

    public UseCase(Scheduler threadExecutor, PostExecutionThread postExecutionThread) {
        this.threadExecutor = threadExecutor;
        this.mainThread = postExecutionThread;
    }

    public abstract Observable<T> buildObservable();

    public Observable<T> execute() {
        return buildObservable()
                .observeOn(mainThread.getScheduler())
                .subscribeOn(threadExecutor);
    }
}

File: app/src/main/java/com/estebanlamas/planesradar/domain/UseCase.kt
package com.estebanlamas.planesradar.domain

import com.estebanlamas.planesradar.domain.executor.PostExecutionThread

import rx.Observable
import rx.Scheduler

/**
 * Created by esteban on 26/12/17
 */

abstract class UseCase<T>(
        private val threadExecutor: Scheduler,
        private val mainThread: PostExecutionThread) {

    abstract fun buildObservable(): Observable<T>

    fun execute(): Observable<T> {
        return buildObservable()
                .observeOn(mainThread.scheduler)
                .subscribeOn(threadExecutor)
    }
}
